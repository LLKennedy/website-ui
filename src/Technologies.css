.Technologies {
  position: relative;
  min-height: 100vh;
  max-height: 100vh;
}

.Technologies-canvas {
  display: block;
  min-height: 100vh;
  max-height: 100vh;
  min-width: 100%;
  max-width: 100%;
}

.Technologies-overlay {
  position: absolute;
  top: 0px;
  left: 0px;
}

.Technologies-textwrapper {
  min-height: 80vh;
  max-height: 80vh;
  max-width: 100vw;
  text-align: center;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
}

.Technologies-textwrapper-bottom {
  min-height: 20vh;
  max-height: 20vh;
  max-width: 100vw;
  text-align: center;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
}

.Technologies-icons {
  max-width: 100vw;
  display: flex;
  align-items: center;
  flex-direction: row;
  justify-content: space-evenly;
}

.Technologies-img {
  max-height: 10vh;
  max-width: 10vh;
  margin-left: 0.5vh;
  margin-right: 0.5vh;
}


.Technologies-text {
  font-size: calc(10px + 2vmin);
  font-size: calc(10px + 2vmin);
  color: rgb(0, 183, 255);
  --shadow-color: #00ccff;
  --shadow-color-light: rgb(31, 0, 168);
  text-shadow: -1px -1px 1px var(--shadow-color-light), -1px 1px 1px var(--shadow-color-light), 1px -1px 1px var(--shadow-color-light), 1px 1px 1px var(--shadow-color-light),
    0 0 3px var(--shadow-color-light), 0 0 10px var(--shadow-color-light)
}