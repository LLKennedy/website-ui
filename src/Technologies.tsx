import React from 'react';
import './Technologies.css'

export interface TechnologiesProps {
    horizontalBars: number;
    verticalBars: number;
    scrollSpeed: number;
    horizonHeightPercent: number;
}

interface TechnologiesState {
    step: number;
    canvasRef: React.RefObject<HTMLCanvasElement>;
}

export default class Technologies extends React.Component<TechnologiesProps, TechnologiesState> {
    public static defaultProps = {
        horizontalBars: 25,
        verticalBars: 20,
        scrollSpeed: 30,
        horizonHeightPercent: 60
    }
    // ctx: CanvasRenderingContext2D;
    constructor(props: TechnologiesProps | Readonly<TechnologiesProps>) {
        super(props);
        this.state = {
            canvasRef: React.createRef<HTMLCanvasElement>(),
            step: 0
        };
        this.paint = this.paint.bind(this);
    }
    render() {
        // useEffect(() => {
        //     let canvas = this?.state?.canvasRef.current;
        //     let context = canvas?.getContext("2d");
        //     context?.beginPath();
        //     context?.arc(50, 50, 50, 0, 2 * Math.PI);
        //     context?.fill();
        // })
        return (
            <div className="Technologies">
                <canvas ref={this?.state?.canvasRef} className="Technologies-canvas" />
            </div>
        )
    }
    componentDidMount() {
        requestAnimationFrame(this.paint);
    }
    paint() {
        let canvas = this?.state?.canvasRef.current;
        if (canvas === undefined || canvas === null) {
            return
        }
        canvas = canvas as HTMLCanvasElement;
        canvas.width = window.innerWidth;
        canvas.height = window.innerHeight;
        let context = canvas.getContext("2d");
        if (context === null) {
            return
        }
        // context.scale(window.devicePixelRatio, window.devicePixelRatio);
        context.clearRect(0, 0, canvas.width, canvas.height);
        if (context.fillStyle) {
            context.fillStyle = "black";
            context.fillRect(0, 0, canvas.width, canvas.height);
        }
        context.beginPath();
        context.fillStyle = "red";
        context.arc(50, 50, 50, 0, 2 * Math.PI);
        context.fill();
        drawLines(context, canvas.height, canvas.width, this.props, this.state);
        this.setState({
            step: (this.state.step + 1) % this.props.scrollSpeed
        })
        requestAnimationFrame(this.paint);
    }
}

function drawLines(ctx: CanvasRenderingContext2D, height: number, width: number, props: Readonly<TechnologiesProps>, state: Readonly<TechnologiesState>) {
    ctx.lineWidth = 1;
    let upperLim = (props.horizonHeightPercent / 100) * height
    let toBottom = Math.log2(height - upperLim)
    drawHorizontalLine(ctx, upperLim, width, "magenta");
    let frameBars = props.horizontalBars * props.scrollSpeed
    for (let i = 0; i < props.horizontalBars; i++) {
        let stepLineY = Math.pow(2, toBottom * (((state.step + (i * props.scrollSpeed)) % frameBars)) / frameBars)
        drawHorizontalLine(ctx, upperLim + stepLineY, width, "magenta");
    }
}

function drawHorizontalLine(ctx: CanvasRenderingContext2D, height: number, width: number, stroke: string | CanvasGradient | CanvasPattern) {
    ctx.strokeStyle = stroke;
    ctx.beginPath();
    ctx.moveTo(0, height);
    ctx.lineTo(width, height);
    ctx.stroke();
}