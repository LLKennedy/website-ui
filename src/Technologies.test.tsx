import React from 'react';
import { render } from '@testing-library/react';
import Technologies, { paintCanvas, TechnologiesProps, TechnologiesState } from './Technologies';
import { Mock } from 'moq.ts';

it('renders correctly', async () => {
    let techRef = React.createRef<Technologies>();
    const tech = render(<Technologies ref={techRef} />);
    expect(tech).toMatchSnapshot();
    let real = techRef.current;
    expect(real).not.toBe(null);
    expect(real).not.toBe(undefined);
});

it('paints without error correctly', async () => {
    const fakeHeight = 1000;
    const fakeWidth = 10000;
    const circleHeight = fakeHeight / 8;
    let canvas = new Mock<HTMLCanvasElement>();
    let ctx = new Mock<CanvasRenderingContext2D>();
    let props: TechnologiesProps = {
        horizonHeightPercent: 50,
        horizontalBars: 1,
        lineStyle: "cyan",
        scrollSpeed: 30,
        vanishingPointHeightExtraPercent: 10,
        verticalBarGapPercent: 50,
        verticalBars: 1
    };
    let fakeRef = React.createRef<HTMLCanvasElement>();
    let state: TechnologiesState = {
        canvasRef: fakeRef,
        step: 0,
    };
    let grad = new Mock<CanvasGradient>();

    grad.setup(instance => instance.addColorStop(0, "red")).returns();
    grad.setup(instance => instance.addColorStop(1, "orange")).returns();

    const gradObj = grad.object();
    expect(gradObj).not.toBeUndefined();

    ctx.setup(instance => instance.clearRect(0, 0, fakeWidth, fakeHeight)).returns();
    ctx.setup(instance => instance.createLinearGradient(0, 0, 0, circleHeight)).returns(gradObj);
    ctx.setup(instance => instance.beginPath()).returns();
    ctx.setup(instance => instance.arc(fakeWidth / 2, circleHeight, circleHeight, 0, Math.PI * 2)).returns();
    ctx.setup(instance => instance.fill()).returns();
    let validStartCoords = [[4875, 137.5], [4875, 162.5], [4875, 187.5], [4875, 212.5], [0, 500], [0, 501], [5000, 1000]];
    let validEndCoords = [[5125, 137.5], [5125, 162.5], [5125, 187.5], [5125, 212.5], [10000, 500], [10000, 501], [5000, 500]];
    ctx.setup(instance => instance.moveTo).returns((x: number, y: number) => {
        expect(validStartCoords).toContainEqual([x, y]);
    })
    ctx.setup(instance => instance.lineTo).returns((x: number, y: number) => {
        expect(validEndCoords).toContainEqual([x, y]);
    })
    ctx.setup(instance => instance.stroke()).returns();


    canvas.setup(instance => instance.clientHeight).returns(fakeHeight);
    canvas.setup(instance => instance.clientWidth).returns(fakeWidth);
    canvas.setup(instance => instance.getContext("2d")).returns(ctx.object())

    paintCanvas(canvas.object(), props, state);
})