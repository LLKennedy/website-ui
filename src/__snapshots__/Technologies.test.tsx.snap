// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="Technologies"
      >
        <canvas
          class="Technologies-canvas"
        />
        <div
          class="Technologies-overlay"
        >
          <div
            class="Technologies-textwrapper"
          >
            <header
              class="Technologies-text"
            >
              I'm a software developer and systems integrator. I have 4 years of experience designing, implementing, maintaining, operating and upgrading software for its entire lifecycle. These are a few of the technologies I use on a regular basis:
            </header>
            <div
              class="Technologies-icons"
            >
              <a
                href="https://www.gnu.org/software/bash/"
              >
                <img
                  alt="Bash"
                  class="Technologies-img"
                  src="https://www.linuxjournal.com/sites/default/files/styles/850x500/public/nodeimage/story/Gnu-bash-logo_1.png"
                />
              </a>
              <a
                href="https://www.centos.org/"
              >
                <img
                  alt="CentOS"
                  class="Technologies-img"
                  src="https://upload.wikimedia.org/wikipedia/commons/thumb/9/9e/CentOS_Graphical_Symbol.svg/1024px-CentOS_Graphical_Symbol.svg.png"
                />
              </a>
              <a
                href="https://www.consul.io/"
              >
                <img
                  alt="Consul"
                  class="Technologies-img"
                  src="https://i1.wp.com/www.stefreitag.de/wp/wp-content/uploads/2019/01/consul-vertical-color.png?resize=678%2C675&ssl=1"
                />
              </a>
              <a
                href="https://www.docker.com/"
              >
                <img
                  alt="Docker"
                  class="Technologies-img"
                  src="https://www.docker.com/sites/default/files/mono-vertical.png"
                />
              </a>
              <a
                href="https://github.com/"
              >
                <img
                  alt="GitHub"
                  class="Technologies-img"
                  src="https://maxcdn.icons8.com/Share/icon/nolan/logos/github1600.png"
                />
              </a>
              <a
                href="https://gitlab.com"
              >
                <img
                  alt="GitLab"
                  class="Technologies-img"
                  src="https://upload.wikimedia.org/wikipedia/commons/1/18/GitLab_Logo.svg"
                />
              </a>
              <a
                href="https://golang.org/"
              >
                <img
                  alt="Go"
                  class="Technologies-img"
                  src="https://upload.wikimedia.org/wikipedia/commons/thumb/2/23/Go_Logo_Aqua.svg/1200px-Go_Logo_Aqua.svg.png"
                />
              </a>
              <a
                href="https://dotnet.microsoft.com/"
              >
                <img
                  alt=".NET Core"
                  class="Technologies-img"
                  src="https://upload.wikimedia.org/wikipedia/commons/thumb/e/ee/.NET_Core_Logo.svg/1200px-.NET_Core_Logo.svg.png"
                />
              </a>
            </div>
            <div
              class="Technologies-icons"
            >
              <a
                href="https://nodejs.org/"
              >
                <img
                  alt="Node.JS"
                  class="Technologies-img"
                  src="https://upload.wikimedia.org/wikipedia/commons/thumb/d/d9/Node.js_logo.svg/1920px-Node.js_logo.svg.png"
                />
              </a>
              <a
                href="https://www.postgresql.org/"
              >
                <img
                  alt="PostgreSQL"
                  class="Technologies-img"
                  src="https://upload.wikimedia.org/wikipedia/commons/thumb/2/29/Postgresql_elephant.svg/1200px-Postgresql_elephant.svg.png"
                />
              </a>
              <a
                href="https://docs.microsoft.com/en-us/powershell/"
              >
                <img
                  alt="PowerShell"
                  class="Technologies-img"
                  src="https://docs.microsoft.com/en-us/powershell/media/index/powershell_128.svg"
                />
              </a>
              <a
                href="https://reactjs.org/"
              >
                <img
                  alt="React"
                  class="Technologies-img"
                  src="https://upload.wikimedia.org/wikipedia/commons/thumb/a/a7/React-icon.svg/1200px-React-icon.svg.png"
                />
              </a>
              <a
                href="https://travis-ci.com/"
              >
                <img
                  alt="Travis CI"
                  class="Technologies-img"
                  src="https://miro.medium.com/max/501/1*IP7q20j8JJK6dB2Jy1pcig.png"
                />
              </a>
              <a
                href="https://www.typescriptlang.org/"
              >
                <img
                  alt="TypeScript"
                  class="Technologies-img"
                  src="https://upload.wikimedia.org/wikipedia/commons/thumb/a/a6/TypeScript_Logo.png/220px-TypeScript_Logo.png"
                />
              </a>
              <a
                href="https://code.visualstudio.com/"
              >
                <img
                  alt="VS Code"
                  class="Technologies-img"
                  src="https://upload.wikimedia.org/wikipedia/commons/thumb/9/9a/Visual_Studio_Code_1.35_icon.svg/1200px-Visual_Studio_Code_1.35_icon.svg.png"
                />
              </a>
              <a
                href="https://www.microsoft.com/en-us/windows"
              >
                <img
                  alt="Windows"
                  class="Technologies-img"
                  src="https://upload.wikimedia.org/wikipedia/commons/thumb/e/ee/Windows_logo_%E2%80%93_2012_%28dark_blue%29.svg/90px-Windows_logo_%E2%80%93_2012_%28dark_blue%29.svg.png"
                />
              </a>
            </div>
          </div>
          <div
            class="Technologies-textwrapper-bottom"
          >
            <header
              class="Technologies-text"
            >
              Besides my technical skills, I enjoy sci-fi and fantasy novels, the entire audiovisual aesthetic of Synthwave/Vaporwave, and thought-provoking or artistic video games such as Outer Wilds, Riven, KSP and Transistor.
            </header>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="Technologies"
    >
      <canvas
        class="Technologies-canvas"
      />
      <div
        class="Technologies-overlay"
      >
        <div
          class="Technologies-textwrapper"
        >
          <header
            class="Technologies-text"
          >
            I'm a software developer and systems integrator. I have 4 years of experience designing, implementing, maintaining, operating and upgrading software for its entire lifecycle. These are a few of the technologies I use on a regular basis:
          </header>
          <div
            class="Technologies-icons"
          >
            <a
              href="https://www.gnu.org/software/bash/"
            >
              <img
                alt="Bash"
                class="Technologies-img"
                src="https://www.linuxjournal.com/sites/default/files/styles/850x500/public/nodeimage/story/Gnu-bash-logo_1.png"
              />
            </a>
            <a
              href="https://www.centos.org/"
            >
              <img
                alt="CentOS"
                class="Technologies-img"
                src="https://upload.wikimedia.org/wikipedia/commons/thumb/9/9e/CentOS_Graphical_Symbol.svg/1024px-CentOS_Graphical_Symbol.svg.png"
              />
            </a>
            <a
              href="https://www.consul.io/"
            >
              <img
                alt="Consul"
                class="Technologies-img"
                src="https://i1.wp.com/www.stefreitag.de/wp/wp-content/uploads/2019/01/consul-vertical-color.png?resize=678%2C675&ssl=1"
              />
            </a>
            <a
              href="https://www.docker.com/"
            >
              <img
                alt="Docker"
                class="Technologies-img"
                src="https://www.docker.com/sites/default/files/mono-vertical.png"
              />
            </a>
            <a
              href="https://github.com/"
            >
              <img
                alt="GitHub"
                class="Technologies-img"
                src="https://maxcdn.icons8.com/Share/icon/nolan/logos/github1600.png"
              />
            </a>
            <a
              href="https://gitlab.com"
            >
              <img
                alt="GitLab"
                class="Technologies-img"
                src="https://upload.wikimedia.org/wikipedia/commons/1/18/GitLab_Logo.svg"
              />
            </a>
            <a
              href="https://golang.org/"
            >
              <img
                alt="Go"
                class="Technologies-img"
                src="https://upload.wikimedia.org/wikipedia/commons/thumb/2/23/Go_Logo_Aqua.svg/1200px-Go_Logo_Aqua.svg.png"
              />
            </a>
            <a
              href="https://dotnet.microsoft.com/"
            >
              <img
                alt=".NET Core"
                class="Technologies-img"
                src="https://upload.wikimedia.org/wikipedia/commons/thumb/e/ee/.NET_Core_Logo.svg/1200px-.NET_Core_Logo.svg.png"
              />
            </a>
          </div>
          <div
            class="Technologies-icons"
          >
            <a
              href="https://nodejs.org/"
            >
              <img
                alt="Node.JS"
                class="Technologies-img"
                src="https://upload.wikimedia.org/wikipedia/commons/thumb/d/d9/Node.js_logo.svg/1920px-Node.js_logo.svg.png"
              />
            </a>
            <a
              href="https://www.postgresql.org/"
            >
              <img
                alt="PostgreSQL"
                class="Technologies-img"
                src="https://upload.wikimedia.org/wikipedia/commons/thumb/2/29/Postgresql_elephant.svg/1200px-Postgresql_elephant.svg.png"
              />
            </a>
            <a
              href="https://docs.microsoft.com/en-us/powershell/"
            >
              <img
                alt="PowerShell"
                class="Technologies-img"
                src="https://docs.microsoft.com/en-us/powershell/media/index/powershell_128.svg"
              />
            </a>
            <a
              href="https://reactjs.org/"
            >
              <img
                alt="React"
                class="Technologies-img"
                src="https://upload.wikimedia.org/wikipedia/commons/thumb/a/a7/React-icon.svg/1200px-React-icon.svg.png"
              />
            </a>
            <a
              href="https://travis-ci.com/"
            >
              <img
                alt="Travis CI"
                class="Technologies-img"
                src="https://miro.medium.com/max/501/1*IP7q20j8JJK6dB2Jy1pcig.png"
              />
            </a>
            <a
              href="https://www.typescriptlang.org/"
            >
              <img
                alt="TypeScript"
                class="Technologies-img"
                src="https://upload.wikimedia.org/wikipedia/commons/thumb/a/a6/TypeScript_Logo.png/220px-TypeScript_Logo.png"
              />
            </a>
            <a
              href="https://code.visualstudio.com/"
            >
              <img
                alt="VS Code"
                class="Technologies-img"
                src="https://upload.wikimedia.org/wikipedia/commons/thumb/9/9a/Visual_Studio_Code_1.35_icon.svg/1200px-Visual_Studio_Code_1.35_icon.svg.png"
              />
            </a>
            <a
              href="https://www.microsoft.com/en-us/windows"
            >
              <img
                alt="Windows"
                class="Technologies-img"
                src="https://upload.wikimedia.org/wikipedia/commons/thumb/e/ee/Windows_logo_%E2%80%93_2012_%28dark_blue%29.svg/90px-Windows_logo_%E2%80%93_2012_%28dark_blue%29.svg.png"
              />
            </a>
          </div>
        </div>
        <div
          class="Technologies-textwrapper-bottom"
        >
          <header
            class="Technologies-text"
          >
            Besides my technical skills, I enjoy sci-fi and fantasy novels, the entire audiovisual aesthetic of Synthwave/Vaporwave, and thought-provoking or artistic video games such as Outer Wilds, Riven, KSP and Transistor.
          </header>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
